### PR TITLE
Implement Optional Attendees

### PR DESCRIPTION
**Pull request by BenLin00**
Optional Attendees considered.
Cleaned up implementation.
Fin.

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -27,10 +27,33 @@
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        List<TimeRange> mandatoryAvailableRanges = new ArrayList<>();
+        List<TimeRange> optionalAttendeesConsideredRanges = new ArrayList<>();
+
+        /* try to remove the timeRanges optional attendees can't make it to. If there's nothing left, ignore the optional attendees*/
+        mandatoryAvailableRanges = availableRanges(events, request, false);
```
Comment (jonathanjlin) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:34]: If you perform some of the checks earlier, you can avoid doing work that you don't need to:
```
List<TimeRange> optionalAttendeesConsideredRanges = availableRanges(events, request, true);
if (request.getAttendees().isEmpty()) {
  return optionalAttendeesConsideredRanges;
}

List<TimeRange> mandatoryAvailableRanges = availableRanges(events, request, false);
...
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -27,10 +27,33 @@
...
+        optionalAttendeesConsideredRanges = availableRanges(events, request, true);
+
+        if (request.getAttendees().isEmpty() ) { // only optional Attendees
+            return optionalAttendeesConsideredRanges;
+        } else if (optionalAttendeesConsideredRanges.isEmpty() && !request.getOptionalAttendees().isEmpty()) { // consider optional Attendees but considering optionalAttendees create no options
+            return mandatoryAvailableRanges;
+        } else {
+            return optionalAttendeesConsideredRanges; // last case: there are optionalAttendees to consider and we can consider optionalAttendees + still have avaliabilities
```
Comment (jonathanjlin) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:42]: typo: availabilities

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -27,10 +27,33 @@
...
         availableRanges.add(TimeRange.WHOLE_DAY);
 
-        Collection<TimeRange> bookedRanges = combinedMandatoryRanges(events, request);
+        Collection<TimeRange> bookedRanges;
+
+        if (considerOptionalAttendees) {
+            bookedRanges = combinedRanges(events, request, true);
+        } else {
+            bookedRanges = combinedRanges(events, request, false);
+        }
```
Comment (jonathanjlin) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:56]: You can simplify this into `Collection<TimeRange> bookedRanges = combinedRanges(events, request, considerOptionalAttendees);`

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -43,56 +66,77 @@
...
+                if (!Collections.disjoint(event.getAttendees(), request.getAttendees()) || //events with mandatory attendees 
+                    !Collections.disjoint(event.getAttendees(), request.getOptionalAttendees())) { // or events with optional attendees
+                    busyTimes.add(event.getWhen());
+                }
+            }
+        } else { // combined events including optional attendees
+            for (Event event : events) {
+                if (!Collections.disjoint(event.getAttendees(), request.getAttendees()) ) { //events with mandatory attendees only
+                    busyTimes.add(event.getWhen());
+                }
```
Comment (jonathanjlin) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:99]: It'll be clearer that the difference between the two is the set of attendees considered if you keep the rest similar, something like:
```
Set<Attendee> attendees = new HashSet<>(request.getAttendees());
if (considerOptionalAttendees) {
  attendees.addAll(request.getOptionalAttendees());
}
for (Event event : events) {
  ...
}
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -43,56 +66,77 @@
...
             }
         }
+
         Collections.sort(busyTimes, TimeRange.ORDER_BY_START);
 
         List<TimeRange> combinedBusyTimes = new ArrayList<>();
 
         int start = -1;
+        int latestEnd = 0;
+        int end = 0;
```
Comment (jonathanjlin) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:109]: If these changes fix a bug, there should be a unit test to verify the case that gets fixed by these changes

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (BenLin00) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:42]: fixed

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (BenLin00) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:56]: Thanks

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (BenLin00) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:34]: Thanks

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (BenLin00) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:99]: Great note! Implemented!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (BenLin00) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:99]: Unit test added and patched hidden bug

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -27,13 +28,28 @@
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        List<TimeRange> optionalAttendeesConsideredRanges = new ArrayList<>();
+
+        /* try to remove the timeRanges optional attendees can't make it to. If there's nothing left, ignore the optional attendees*/
+        optionalAttendeesConsideredRanges = availableRanges(events, request, true);
```
Comment (jonathanjlin) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:34]: Combine variable declaration and initialization unless you need to separate them: `List<TR> oACR = availableRanges(events, request, true);`

Additionally, for primitive type parameters (String, int, boolean, etc), add a comment to help readers quickly understand what it's for. Within Google the convention is `availableRanges(events, requests, /* considerOptionalAttendees= */ true);`

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -27,13 +28,28 @@
...
+
+        if (request.getAttendees().isEmpty() ) { // only optional Attendees
```
Comment (jonathanjlin) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:36]: I think the ordering of the logic here is a bit hard to follow: intuitively I would expect that the first check would be:
```
if (!optionalAttendeesConsideredRanges.isEmpty()) {
  return optionalAttendeesConsideredRanges;
}
```
since if there's times available for both required and optional attendees, we should return that.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 6 above)
Comment (BenLin00) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:34]: noted and implemented!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (BenLin00) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: resolved
